src/main/java/org/jenkinsci/maven/plugins/hpi/RunMojo.java

```diff
@@ -783,22 +801,39 @@ public void startJetty() throws MojoExecutionException {
      */
     protected Set<Artifact> resolveDependencies(String scope) throws MojoExecutionException {
         try {
-            ArtifactResolutionResult result = artifactResolver.resolveTransitively(
-                    getProject().getDependencyArtifacts(),
-                    getProject().getArtifact(),
-                    getProject().getManagedVersionMap(),
-                    localRepository,
-                    getProject().getRemoteArtifactRepositories(),
-                    artifactMetadataSource,
-                    new ScopeArtifactFilter(scope));
-            return result.getArtifacts();
-        } catch (ArtifactNotFoundException e) {
-            throw new MojoExecutionException("Unable to copy dependency plugin",e);
-        } catch (ArtifactResolutionException e) {
+            DefaultDependencyResolutionRequest request =
+                    new DefaultDependencyResolutionRequest(project, session.getRepositorySession());
+            request.setResolutionFilter(getDependencyFilter(scope));
+            DependencyResolutionResult result = dependenciesResolver.resolve(request);
+
+            Set<Artifact> artifacts = new LinkedHashSet<>();
+            if (result.getDependencyGraph() != null
+                    && !result.getDependencyGraph().getChildren().isEmpty()) {
+                RepositoryUtils.toArtifacts(
+                        artifacts,
+                        result.getDependencyGraph().getChildren(),
+                        Collections.singletonList(project.getArtifact().getId()),
+                        request.getResolutionFilter());
+            }
```

Comment:
Did you find what was wrong with the maven 3 way of doing things? Bug on maven side? It looks weird that the dependency trail wouldn't be populated without this.

Reply:
>the maven 3 way of doing things

Does not make sense. The comment to `RunMojo#resolveDependencies` says "Performs the equivalent of "@requireDependencyResolution" mojo attribute", and the implementation of `requiresDependencyResolution` in `LifecycleDependencyResolver` uses `ProjectDependenciesResolver`, so my revised code (_not_ the original) _is_ the "Maven 3 way of doing things".